src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php

```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
         $this->compiler = $container->getCompiler();
         $this->formatter = $this->compiler->getLoggingFormatter();
 
-        foreach ($container->getAliases() as $id => $alias) {
+        $seenAliases = array();
+        $replacements = array();
+        foreach ($container->getAliases() as $defId => $alias) {
             $aliasId = (string) $alias;
-
+            if (isset($replacements[$aliasId])) {
+                $container->setAlias($defId, $replacements[$aliasId]);
```

Comment:
I tried to stick to the original behaviour, namely: remove the definition of the target, point the alias to the definition, and point every subsequent alias with the same target to the first alias.